camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/common/KafkaReader.java

```diff
@@ -163,53 +168,91 @@ public boolean fetch() throws IOException {
+    try {
+      refreshTopicMetadata();
+      FetchResponse fetchResponse = simpleConsumer.fetch(fetchRequest);
+      if (fetchResponse.hasError()) {
         log.info("Error encountered during a fetch request from Kafka");
         log.info("Error Code generated : "
             + fetchResponse.errorCode(kafkaRequest.getTopic(), kafkaRequest.getPartition()));
         return false;
-      } else {
-        ByteBufferMessageSet messageBuffer =
-            fetchResponse.messageSet(kafkaRequest.getTopic(), kafkaRequest.getPartition());
-        lastFetchTime = (System.currentTimeMillis() - tempTime);
-        log.debug("Time taken to fetch : " + (lastFetchTime / 1000) + " seconds");
-        log.debug("The size of the ByteBufferMessageSet returned is : " + messageBuffer.sizeInBytes());
-        int skipped = 0;
-        totalFetchTime += lastFetchTime;
-        messageIter = messageBuffer.iterator();
-        //boolean flag = false;
-        Iterator<MessageAndOffset> messageIter2 = messageBuffer.iterator();
-        MessageAndOffset message = null;
-        while (messageIter2.hasNext()) {
-          message = messageIter2.next();
-          if (message.offset() < currentOffset) {
-            //flag = true;
-            skipped++;
-          } else {
-            log.debug("Skipped offsets till : " + message.offset());
-            break;
-          }
-        }
-        log.debug("Number of offsets to be skipped: " + skipped);
-        while (skipped != 0) {
-          MessageAndOffset skippedMessage = messageIter.next();
-          log.debug("Skipping offset : " + skippedMessage.offset());
-          skipped--;
-        }
+      }
+      return processFetchResponse(fetchResponse, tempTime);
+    } catch (Exception e) {
+      log.info("Exception generated during fetch for topic " + kafkaRequest.getTopic()
+          + ". This topic will be skipped.");
+      e.printStackTrace();
+      return false;
+    }
+  }
 
-        if (!messageIter.hasNext()) {
-          System.out.println("No more data left to process. Returning false");
-          messageIter = null;
-          return false;
+  private void refreshTopicMetadata() {
+    TopicMetadataRequest request = new TopicMetadataRequest(Collections.singletonList(kafkaRequest.getTopic()));
+    TopicMetadataResponse response = simpleConsumer.send(request);
+    TopicMetadata metadata = response.topicsMetadata().get(0);
+    for (PartitionMetadata partitionMetadata : metadata.partitionsMetadata()) {
+      if (partitionMetadata.partitionId() == kafkaRequest.getPartition()) {
+        simpleConsumer = new SimpleConsumer(partitionMetadata.leader().host(), partitionMetadata.leader().port(),
+            CamusJob.getKafkaTimeoutValue(context), CamusJob.getKafkaBufferSize(context),
+            CamusJob.getKafkaClientName(context));
+        break;
+      }
+    }
+  }
+
+  private boolean processFetchResponse(FetchResponse fetchResponse, long tempTime) {
```

Comment:
Line 215-249 is the same as old code line 171-205. I added a try-catch outside it, because there's a try catch in old code 207-211.